File: DSA/Arrays/Code/9_SearchingInArrays.cs
using System;

// Searching in Arrays in C#
// Linear Search and Binary Search implementations

class SearchingInArrays {
    
    // Linear Search
    static int LinearSearch(int[] arr, int target) {
        for (int i = 0; i < arr.Length; i++) {
            if (arr[i] == target)
                return i;
        }
        return -1;
    }
    
    // Binary Search
    static int BinarySearch(int[] arr, int target) {
        int left = 0, right = arr.Length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (arr[mid] == target)
                return mid;
            else if (arr[mid] < target)
                left = mid + 1;
            else
                right = mid - 1;
        }
        return -1;
    }
    
    static void Main() {
        int[] arr = {10, 20, 30, 40, 50, 60, 70};
        int target = 40;
        
        // Linear Search
        Console.Write("Linear Search for " + target + ": ");
        int pos = LinearSearch(arr, target);
        if (pos != -1)
            Console.WriteLine("Found at index " + pos);
        else
            Console.WriteLine("Not found");
        
        // Linear Search for non-existing element
        target = 25;
        Console.Write("Linear Search for " + target + ": ");
        pos = LinearSearch(arr, target);
        if (pos != -1)
            Console.WriteLine("Found at index " + pos);
        else
            Console.WriteLine("Not found");
        
        // Binary Search
        target = 50;
        Console.Write("Binary Search for " + target + ": ");
        pos = BinarySearch(arr, target);
        if (pos != -1)
            Console.WriteLine("Found at index " + pos);
        else
            Console.WriteLine("Not found");
        
        // Using C#'s built-in method
        target = 30;
        Console.Write("Using Array.BinarySearch for " + target + ": ");
        pos = Array.BinarySearch(arr, target);
        if (pos >= 0)
            Console.WriteLine("Found at index " + pos);
        else
            Console.WriteLine("Not found");
    }
}

File: DSA/Arrays/Code/9_SearchingInArrays.c
#include <stdio.h>

// Searching in Arrays in C
// Linear Search and Binary Search implementations

// Linear Search
int linearSearch(int arr[], int n, int target) {
    for (int i = 0; i < n; i++) {
        if (arr[i] == target)
            return i;
    }
    return -1;
}

// Binary Search (for sorted array)
int binarySearch(int arr[], int n, int target) {
    int left = 0, right = n - 1;
    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (arr[mid] == target)
            return mid;
        else if (arr[mid] < target)
            left = mid + 1;
        else
            right = mid - 1;
    }
    return -1;
}

int main() {
    int arr[] = {10, 20, 30, 40, 50, 60, 70};
    int n = 7;
    int target = 40;
    
    // Linear Search
    printf("Linear Search for %d: ", target);
    int pos = linearSearch(arr, n, target);
    if (pos != -1)
        printf("Found at index %d\n", pos);
    else
        printf("Not found\n");
    
    // Linear Search for non-existing element
    target = 25;
    printf("Linear Search for %d: ", target);
    pos = linearSearch(arr, n, target);
    if (pos != -1)
        printf("Found at index %d\n", pos);
    else
        printf("Not found\n");
    
    // Binary Search (array is sorted)
    target = 50;
    printf("Binary Search for %d: ", target);
    pos = binarySearch(arr, n, target);
    if (pos != -1)
        printf("Found at index %d\n", pos);
    else
        printf("Not found\n");
    
    return 0;
}
